File: services/feed-ingester/internal/handler/livescore.go
package handler

import (
	"encoding/json"
	"time"

	"github.com/oddzilla/feed-ingester/internal/oddinxml"
)

// liveScorePayload is the JSON shape persisted to matches.live_score and
// returned verbatim by the API. The top-level `home`/`away` fields hold
// the series score (number of maps won) so existing list-card code that
// reads `liveScore.home` / `liveScore.away` keeps working unchanged.
//
// Per-map detail (rounds for CS2/Valorant, kills for Dota/LoL, etc.) is
// in `periods`; the live state of the in-progress map is in `scoreboard`.
// `currentMap` is 1-indexed and reflects which map is being played right
// now — derived from completed-period count or, when unavailable, from
// home+away map score.
type liveScorePayload struct {
	Home            *int               `json:"home,omitempty"`
	Away            *int               `json:"away,omitempty"`
	Status          *int               `json:"status,omitempty"`
	MatchStatusCode *int               `json:"matchStatusCode,omitempty"`
	CurrentMap      *int               `json:"currentMap,omitempty"`
	Scoreboard      *scoreboardPayload `json:"scoreboard,omitempty"`
	Periods         []periodPayload    `json:"periods,omitempty"`
	UpdatedAt       string             `json:"updatedAt"`
}

type scoreboardPayload struct {
	HomeWonRounds        *int   `json:"homeWonRounds,omitempty"`
	AwayWonRounds        *int   `json:"awayWonRounds,omitempty"`
	HomeKills            *int   `json:"homeKills,omitempty"`
	AwayKills            *int   `json:"awayKills,omitempty"`
	HomeDestroyedTurrets *int   `json:"homeDestroyedTurrets,omitempty"`
	AwayDestroyedTurrets *int   `json:"awayDestroyedTurrets,omitempty"`
	HomeDestroyedTowers  *int   `json:"homeDestroyedTowers,omitempty"`
	AwayDestroyedTowers  *int   `json:"awayDestroyedTowers,omitempty"`
	HomeGold             *int   `json:"homeGold,omitempty"`
	AwayGold             *int   `json:"awayGold,omitempty"`
	HomeGoals            *int   `json:"homeGoals,omitempty"`
	AwayGoals            *int   `json:"awayGoals,omitempty"`
	CurrentCtTeam        *int   `json:"currentCtTeam,omitempty"`
	CurrentDefTeam       *int   `json:"currentDefTeam,omitempty"`
	Time                 string `json:"time,omitempty"`
	GameTime             *int   `json:"gameTime,omitempty"`
	RemainingGameTime    *int   `json:"remainingGameTime,omitempty"`
}

type periodPayload struct {
	Number               *int   `json:"number,omitempty"`
	Type                 string `json:"type,omitempty"`
	MatchStatusCode      *int   `json:"matchStatusCode,omitempty"`
	HomeScore            *int   `json:"homeScore,omitempty"`
	AwayScore            *int   `json:"awayScore,omitempty"`
	HomeWonRounds        *int   `json:"homeWonRounds,omitempty"`
	AwayWonRounds        *int   `json:"awayWonRounds,omitempty"`
	HomeKills            *int   `json:"homeKills,omitempty"`
	AwayKills            *int   `json:"awayKills,omitempty"`
	HomeGoals            *int   `json:"homeGoals,omitempty"`
	AwayGoals            *int   `json:"awayGoals,omitempty"`
	HomeDestroyedTurrets *int   `json:"homeDestroyedTurrets,omitempty"`
	AwayDestroyedTurrets *int   `json:"awayDestroyedTurrets,omitempty"`
	HomeDestroyedTowers  *int   `json:"homeDestroyedTowers,omitempty"`
	AwayDestroyedTowers  *int   `json:"awayDestroyedTowers,omitempty"`
	IsLive               bool   `json:"isLive,omitempty"`
}

// buildLiveScore converts a decoded sport_event_status into the JSON
// payload we persist. Returns (nil, nil) when the block has nothing
// useful (no series score, no periods, no scoreboard) — caller should
// then leave matches.live_score as-is.
//
// `currentMap` heuristic:
//   - Prefer the highest period.number with match_status_code that maps
//     to "in progress" (Oddin uses 6 = live for CS2 maps).
//   - Otherwise: count completed periods + 1 (next map being played).
//   - Otherwise: home+away series score + 1 (e.g. 2-1 series → map 4).
//   - The match must be live; for closed matches the indicator is dropped.
func buildLiveScore(s *oddinxml.SportEventStatus, oddinTsMs int64) ([]byte, error) {
	if s == nil {
		return nil, nil
	}
	hasPeriods := s.PeriodScores != nil && len(s.PeriodScores.Periods) > 0
	hasScoreboard := s.Scoreboard != nil
	if s.HomeScore == nil && s.AwayScore == nil && !hasPeriods && !hasScoreboard {
		return nil, nil
	}

	payload := liveScorePayload{
		Home:            s.HomeScore,
		Away:            s.AwayScore,
		Status:          s.Status,
		MatchStatusCode: s.MatchStatus,
		UpdatedAt:       time.UnixMilli(oddinTsMs).UTC().Format(time.RFC3339Nano),
	}

	if hasScoreboard {
		sb := s.Scoreboard
		payload.Scoreboard = &scoreboardPayload{
			HomeWonRounds:        sb.HomeWonRounds,
			AwayWonRounds:        sb.AwayWonRounds,
			HomeKills:            sb.HomeKills,
			AwayKills:            sb.AwayKills,
			HomeDestroyedTurrets: sb.HomeDestroyedTurrets,
			AwayDestroyedTurrets: sb.AwayDestroyedTurrets,
			HomeDestroyedTowers:  sb.HomeDestroyedTowers,
			AwayDestroyedTowers:  sb.AwayDestroyedTowers,
			HomeGold:             sb.HomeGold,
			AwayGold:             sb.AwayGold,
			HomeGoals:            sb.HomeGoals,
			AwayGoals:            sb.AwayGoals,
			CurrentCtTeam:        sb.CurrentCtTeam,
			CurrentDefTeam:       sb.CurrentDefTeam,
			Time:                 sb.Time,
			GameTime:             sb.GameTime,
			RemainingGameTime:    sb.RemainingGameTime,
		}
	}

	if hasPeriods {
		periods := s.PeriodScores.Periods
		payload.Periods = make([]periodPayload, 0, len(periods))
		for _, p := range periods {
			payload.Periods = append(payload.Periods, periodPayload{
				Number:               p.Number,
				Type:                 p.Type,
				MatchStatusCode:      p.MatchStatusCode,
				HomeScore:            p.HomeScore,
				AwayScore:            p.AwayScore,
				HomeWonRounds:        p.HomeWonRounds,
				AwayWonRounds:        p.AwayWonRounds,
				HomeKills:            p.HomeKills,
				AwayKills:            p.AwayKills,
				HomeGoals:            p.HomeGoals,
				AwayGoals:            p.AwayGoals,
				HomeDestroyedTurrets: p.HomeDestroyedTurrets,
				AwayDestroyedTurrets: p.AwayDestroyedTurrets,
				HomeDestroyedTowers:  p.HomeDestroyedTowers,
				AwayDestroyedTowers:  p.AwayDestroyedTowers,
			})
		}
	}

	payload.CurrentMap = deriveCurrentMap(s)

	// Mark the live period so the frontend can highlight it without
	// re-deriving the index. A period is live when its number matches
	// `currentMap` AND the overall match is still running (status<3 in
	// UOF terms — anything ≥3 is ended/closed/cancelled).
	if payload.CurrentMap != nil && (s.Status == nil || *s.Status < 3) {
		for i := range payload.Periods {
			if payload.Periods[i].Number != nil && *payload.Periods[i].Number == *payload.CurrentMap {
				payload.Periods[i].IsLive = true
			}
		}
	}

	return json.Marshal(payload)
}

func deriveCurrentMap(s *oddinxml.SportEventStatus) *int {
	if s == nil {
		return nil
	}
	// If the series is over, no map is "live".
	if s.Status != nil && *s.Status >= 3 {
		return nil
	}

	// Prefer the explicit per-period in-progress flag. Oddin uses
	// match_status_code 6 ("In progress") on the live map; values >= 100
	// indicate the period is finished. We pick the highest live number.
	if s.PeriodScores != nil {
		var live, maxFinished int
		var liveSet, finishedSet bool
		for _, p := range s.PeriodScores.Periods {
			if p.Number == nil {
				continue
			}
			if p.MatchStatusCode != nil && *p.MatchStatusCode == 6 {
				if !liveSet || *p.Number > live {
					live = *p.Number
					liveSet = true
				}
				continue
			}
			if !finishedSet || *p.Number > maxFinished {
				maxFinished = *p.Number
				finishedSet = true
			}
		}
		if liveSet {
			n := live
			return &n
		}
		if finishedSet {
			n := maxFinished + 1
			return &n
		}
	}

	if s.HomeScore != nil && s.AwayScore != nil {
		n := *s.HomeScore + *s.AwayScore + 1
		return &n
	}
	return nil
}

File: apps/web/src/lib/live-score.ts
// Shared types + helpers for the live-score JSON shape feed-ingester
// writes to matches.live_score (see services/feed-ingester/internal/
// handler/livescore.go). Used by both the match-detail Scoreboard and
// the match-list row's compact per-map mini-table so the two views
// stay byte-identical on which metric appears in each cell.

export interface LiveScorePeriod {
  number?: number;
  type?: string;
  matchStatusCode?: number;
  homeScore?: number;
  awayScore?: number;
  homeWonRounds?: number;
  awayWonRounds?: number;
  homeKills?: number;
  awayKills?: number;
  homeGoals?: number;
  awayGoals?: number;
  homeDestroyedTurrets?: number;
  awayDestroyedTurrets?: number;
  homeDestroyedTowers?: number;
  awayDestroyedTowers?: number;
  isLive?: boolean;
}

export interface LiveScoreScoreboard {
  homeWonRounds?: number;
  awayWonRounds?: number;
  homeKills?: number;
  awayKills?: number;
  homeDestroyedTurrets?: number;
  awayDestroyedTurrets?: number;
  homeDestroyedTowers?: number;
  awayDestroyedTowers?: number;
  homeGold?: number;
  awayGold?: number;
  homeGoals?: number;
  awayGoals?: number;
  currentCtTeam?: number;
  currentDefTeam?: number;
  time?: string;
  gameTime?: number;
  remainingGameTime?: number;
}

/**
 * The match clock as an ANCHOR rather than a reading: at unix-ms
 * instant `atMs` (the feed's server time) the clock read `seconds`, and
 * it has moved `direction` seconds per second since — 1 counting up
 * (football, hockey, basketball, esports on the Fonbet line), 0 stopped
 * (half time, an intermission, "END"), -1 counting down. The reading at
 * any instant t is `seconds + direction × (t − atMs) / 1000`; see
 * `clockSecondsAt` in clock-math.ts, and `useRunningClock` for the hook
 * that ticks it.
 *
 * Written by fonbet-ingester, normalised so a RUNNING clock is stored at
 * its zero instant (`seconds` 0, `atMs` = when it read 0) — a form that
 * is constant for as long as the clock runs, so the payload does not
 * change from poll to poll and a single frame carries a whole half. A
 * stopped clock keeps its reading in `seconds` and has no `atMs`. Absent
 * on the Oddin esports payload and on untimed sports (tennis, table
 * tennis, volleyball, cricket, snooker), where `scoreboard.time` — the
 * feed's own rendered string, static between frames — is the fallback.
 */
export interface LiveScoreClock {
  seconds: number;
  direction: number;
  atMs?: number;
}

export interface LiveScore {
  home?: number;
  away?: number;
  status?: number;
  matchStatusCode?: number;
  currentMap?: number;
  clock?: LiveScoreClock | null;
  /**
   * Which side holds serve: 1 = home, 2 = away. Written by
   * fonbet-ingester for the sports whose feed carries it (tennis,
   * table tennis, volleyball) and absent everywhere else — Oddin's
   * esports payload has no equivalent. Read through `servingSide()`
   * rather than compared inline, so an unexpected value degrades to
   * "nobody is marked" instead of marking the home player.
   */
  serve?: number;
  /**
   * The feed's own parenthetical qualifier on the headline score —
   * games in the current set for tennis ("(6-5)"), the period line
   * elsewhere. Written by fonbet-ingester from Fonbet's `comment` /
   * `scoreComment`; absent on the Oddin esports payload, where the
   * headline score IS the map count and there is nothing to qualify.
   *
   * Rendered verbatim rather than reconstructed from `periods`: it is
   * the string Fonbet itself shows, and it already encodes per-sport
   * conventions we would otherwise have to re-derive per sport.
   */
  comment?: string;
  scoreboard?: LiveScoreScoreboard;
  periods?: LiveScorePeriod[];
  updatedAt?: string;
}

// servingSide answers "is this side serving right now" for one row.
// Only meaningful while the match is live: the feed keeps the last
// value on the payload after a match ends, and a serve marker on a
// finished scoreboard reads as a live match that isn't.
export function servingSide(
  liveScore: LiveScore | null | undefined,
  isLive: boolean,
): "home" | "away" | null {
  if (!isLive) return null;
  if (liveScore?.serve === 1) return "home";
  if (liveScore?.serve === 2) return "away";
  return null;
}

// mapCellValue picks the right metric for one (team, map) cell.
// Prefers the live <scoreboard> for the current map (more up-to-date than
// the period_score row, which only refreshes after the map ends on some
// sports). Falls back to the period when the scoreboard block is empty —
// Oddin sometimes ships `<scoreboard/>` with no inner attributes during
// pauses or just-started maps.
export function mapCellValue(
  side: "home" | "away",
  mapNumber: number,
  period: LiveScorePeriod | undefined,
  scoreboard: LiveScoreScoreboard | null,
  currentMap: number | null,
  sportSlug: string,
): number | null {
  if (currentMap === mapNumber && scoreboard) {
    const live = pickLiveMetric(scoreboard, sportSlug);
    const v = side === "home" ? live.home : live.away;
    if (v != null) return v;
  }
  if (!period) return null;
  const periodMetric = pickPeriodMetric(period, sportSlug);
  return side === "home" ? periodMetric.home : periodMetric.away;
}

// pickPeriodMetric returns the (home, away) integer pair for a completed
// map, preferring the sport-native metric (rounds for CS2, kills for Dota)
// and falling back to the generic home_score/away_score Oddin always
// populates.
export function pickPeriodMetric(
  p: LiveScorePeriod,
  sportSlug: string,
): { home: number | null; away: number | null } {
  const slug = sportSlug.toLowerCase();
  if (slug === "cs2" || slug === "csgo" || slug === "counter-strike" || slug === "valorant") {
    if (p.homeWonRounds != null || p.awayWonRounds != null) {
      return { home: p.homeWonRounds ?? null, away: p.awayWonRounds ?? null };
    }
  }
  if (
    slug === "dota2" ||
    slug === "dota" ||
    slug === "lol" ||
    slug === "leagueoflegends" ||
    slug === "league-of-legends"
  ) {
    if (p.homeKills != null || p.awayKills != null) {
      return { home: p.homeKills ?? null, away: p.awayKills ?? null };
    }
  }
  if (p.homeScore != null || p.awayScore != null) {
    return { home: p.homeScore ?? null, away: p.awayScore ?? null };
  }
  if (p.homeGoals != null || p.awayGoals != null) {
    return { home: p.homeGoals ?? null, away: p.awayGoals ?? null };
  }
  return { home: null, away: null };
}

export function pickLiveMetric(
  sb: LiveScoreScoreboard,
  sportSlug: string,
): { home: number | null; away: number | null } {
  const slug = sportSlug.toLowerCase();
  if (slug === "cs2" || slug === "csgo" || slug === "counter-strike" || slug === "valorant") {
    if (sb.homeWonRounds != null || sb.awayWonRounds != null) {
      return { home: sb.homeWonRounds ?? null, away: sb.awayWonRounds ?? null };
    }
  }
  if (
    slug === "dota2" ||
    slug === "dota" ||
    slug === "lol" ||
    slug === "leagueoflegends" ||
    slug === "league-of-legends"
  ) {
    if (sb.homeKills != null || sb.awayKills != null) {
      return { home: sb.homeKills ?? null, away: sb.awayKills ?? null };
    }
  }
  if (sb.homeGoals != null || sb.awayGoals != null) {
    return { home: sb.homeGoals ?? null, away: sb.awayGoals ?? null };
  }
  return { home: null, away: null };
}
